Orange/canvas/scheme/readwrite.py

```diff
@@ -905,6 +906,165 @@ def indent_(element, level, last):
 
     return indent_(element, level, True)
 
+# Parent nodes are nodes with no inputs
+def get_parent_nodes(links, viable_nodes, adjs):
+    parents = []
+    for node in viable_nodes:
+        parent = True
+        for adj in adjs:
+            if adj[1] == node:
+                parent = False
+                break
+        if parent:
+            parents.append(node)
+    return parents
+
+def get_links_from(parent_node, adjs):
+    matches = []
+    for adj in adjs:
+        if adj[0] == parent_node:
+            matches.append(adj)
+    return matches
+
+def get_links_to(child_node, adjs):
+    matches = []
+    for adj in adjs:
+        if adj[1] == child_node:
+            matches.append(adj)
+    return matches
```

Comment:
Just an idea to shorten this function using list comprehension
```
def get_links_to(child_node, adjs):
        return [adj for adj in adjs if adj[1] == child_node]
```
Same can be done for `get_links_from` function